README.md
# 15 Puzzle

Chat-bot implementation of the classic 15 Puzzle game build on top of the MiniApps plaftorm.


history.txt
Ver-0.0.1: makar
2016-08-05: Alfa version

Ver-0.0.2: makar
2016-12-06: gp_lang parameter defines user language when user's game context is created
2016-12-06: gp_back_url parameter defines back URL for GamePortal

Ver-0.0.2: makar
2016-12-07: gp_back_url fix (&amp; encoding)
2016-12-07: some cosmetic changes (icon & title for "another_games")

test/src/Test.java
/**
 * Created by makarov on 16.06.16.
 */
public class Test
{
  public static void main(String[] args) {

  }
}


test/StoreUtility.java
import mobi.eyeline.utils.restclient.web.RestClient;
import org.apache.log4j.Logger;

import static mobi.eyeline.utils.restclient.web.RestClient.delete;
import static mobi.eyeline.utils.restclient.web.RestClient.post;

import java.io.IOException;
import java.text.DateFormat;
import java.util.Date;

public class StoreUtility
{
  //private final static Logger logger = Logger.getLogger(StoreUtility.class);
  static final String API_ROOT = "http://ec2.globalussd.mobi:11201/wstorage";

  public static void main(String[] args) throws IOException
  {
    final String wnumber = "a761f808-da01-48ea-add8-360bef724af9";
    final String pers_var = "services.game15.pers";
    final String sess_var = "services.game15.sess";

    //new RestClient().json(API_ROOT + "/profile/" + wnumber + "/mobile.msisdn", post(RestClient.content("123456")));
    new RestClient().json(API_ROOT + "/profile/" + wnumber + "/" + pers_var, delete());
    new RestClient().json(API_ROOT + "/profile/" + wnumber + "/" + sess_var, delete());

    String data = null;
    try {
      data = (String)((new RestClient()).json(API_ROOT + "/profile/" + wnumber + "/" + pers_var).object().get("value"));
      if (data != null) {
        System.out.println("Load game for '" + wnumber + "', var=" + pers_var + ", data='" + data + "'. Ok");
      }
    } catch (Exception exc) {
      System.out.println("Load game for '" + wnumber + "', var=" + pers_var + ". Failed. Details: " + exc.getMessage());
    }

    /*final int time = 40*3600 + 3600*15 + 60*10 + 15;
    final StringBuilder sb = new StringBuilder(64);

    int val = time / (3600 * 24);
    if (val > 0) { sb.append(val); sb.append(' '); sb.append("days"); sb.append(' '); }
    val = (time / 3600) % 24; if (val < 10) sb.append('0'); sb.append(val); sb.append(':');
    val = (time / 60) % 60; if (val < 10) sb.append('0'); sb.append(val); sb.append(':');
    val = time % 60; if (val < 10) sb.append('0'); sb.append(val);

    System.out.println("Time: " + sb.toString());*/

    System.out.println("Done");
  }
}


src/mobi/eyeline/bots/game15/Game.java
package mobi.eyeline.bots.game15;

import org.apache.log4j.Logger;

import java.text.DateFormat;
import java.util.*;

public class Game
{
  private final static Logger logger = Logger.getLogger(Game.class);

  public static class Score implements Comparable<Score>
  {
    protected int  time  = 0;
    protected int  moves = 0;

    protected Score(int time, int moves) {
      this.time = time; this.moves = moves;
    }
    protected Score(Score s) {
      this(s.time, s.moves);
    }
    public Score() {
      this(0, 0);
    }

    @Override
    public int compareTo(Score o) {
      return (this.time == o.time) ? this.moves-o.moves : this.time-o.time;
    }

    public synchronized int getTime() {
      return time;
    }
    public synchronized int getMoves() {
      return this.moves;
    }
  }

  public static class GameScore extends Score
  {
    protected long startTime;

    public GameScore() {
      super(); this.startTime = System.currentTimeMillis();
    }
    protected GameScore(int time, int moves) {
      this(); this.startTime -= time*1000;
      this.time = time; this.moves = moves;
    }

    @Override
    public synchronized int getTime() {
      time = (int)(System.currentTimeMillis() - startTime)/1000;
      return time;
    }

    public synchronized int incMoves() {
      this.moves++;
      return this.moves;
    }
  }

  private String    subscriber;
  private int       subscriberLocale = GameResourceBundle.DEFAULT_LOCALE; //pers

  private long      firstTime;            //pers
  private long      totalGamesCount = 0;  //pers
  private long      winGamesCount = 0;    //pers
  private Score     bestScore = null;     //pers
  private boolean   bestScoreSet = false;

  private long      lastTime;             //sess
  private GameScore currScore = null;     //sess
  private byte[][]  cells;                //sess
  private int       empty;                //sess
  private boolean   gameStarted = false;  //sess
  private int       easyDebug = 0; //TODO: remove for fare game

  private GameResourceBundle grb = null;

  public static boolean isSubscriberValid(String subscriber) {
    return subscriber != null; // TODO: Check MSISDN
  }

  public Game(String subscriber) {
    this.subscriber = subscriber;
    this.subscriberLocale = GameResourceBundle.DEFAULT_LOCALE;
    this.grb = new GameResourceBundle(this.subscriberLocale);
    this.firstTime = System.currentTimeMillis();
    this.lastTime = this.firstTime;
  }
  protected Game(String subscriber, String persData) throws Exception {
    this.subscriber = subscriber;
    loadGamePers(persData);
  }
  protected Game(String subscriber, String persData, String sessData) throws Exception {
    this(subscriber, persData);
    loadGameSess(sessData);
  }

  public void setGpLang(String gp_lang) {
    this.subscriberLocale = ("RU".equalsIgnoreCase(gp_lang)) ? GameResourceBundle.RU_RU_LOCALE :
                            ("EN".equalsIgnoreCase(gp_lang)) ? GameResourceBundle.EN_US_LOCALE :
                                                               GameResourceBundle.DEFAULT_LOCALE;
    this.grb = new GameResourceBundle(this.subscriberLocale);
  }

  private String gpBackURL = null;
  public void setGpBackURL(String gpBackURL) {
    this.gpBackURL = gpBackURL;
  }
  public String getGpBackURL() {
    String url = (this.gpBackURL != null) ? this.gpBackURL : GameStore.getAnotherGamesUrl();
    String processedUrl = null;
    if (url != null) processedUrl = url.replace("&", "&amp;");
    return processedUrl;
  }

  public String getSubscriber() {
    return subscriber;
  }
  public void switchLocale() {
    grb.switchLocale();
    easyDebug++;
  }
  public String getText(String key) {
    return grb.getString(key);
  }
  public String getNextText(String key) {
    return grb.getNextString(key);
  }
  public byte[][] getCells() {
    return cells;
  }
  public long getTotalGames() {
    return totalGamesCount;
  }
  public long getWinGames() {
    return winGamesCount;
  }

  private String formatDate(long date) {
    if (date <= 0 || grb == null) return "";
    final DateFormat df = DateFormat.getDateTimeInstance(DateFormat.LONG, DateFormat.MEDIUM, grb.getCurrentLocale());
    return df.format(new Date(date));
  }
  public String getFirstTime() {
    return formatDate(firstTime);
  }
  public String getLastTime() {
    return formatDate(lastTime);
  }

  private void newCells() {
    this.cells = new byte[][] {{0,1,2,3}, {4,5,6,7}, {8,9,10,11}, {12,13,14,15}};
    this.empty = 0;
  }
  public void newGame()
  {
    newCells();

    final Byte[] array = {0, 2, 3, 4, 1, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15};
    final ArrayList<Byte> list = new ArrayList<>(Arrays.asList(array));
    if (easyDebug <= 0 || (easyDebug % 5 != 0))
      Collections.shuffle(list, new Random(System.nanoTime()));

    for (int y=0,i=0; y<4; y++) {
      for (int x=0; x<4; x++,i++) {
        byte b = list.get(i); this.cells[y][x] = b;
        if (b == 0) this.empty = i;
      }
    }

    this.lastTime = System.currentTimeMillis();
    this.totalGamesCount++; this.bestScoreSet = false;
    this.currScore = new GameScore();
    this.gameStarted = true;
  }

  private void loadGamePers(String data) throws Exception
  {
    final String GAME_LOAD_ERROR = "Game load error (pers): ";
    final String PARSE_FAILED = "Failed to parse game data ";

    if (data == null)
      throw new Exception(GAME_LOAD_ERROR + "Game data is null");

    try {
      StringTokenizer st = new StringTokenizer(data, ",");

      this.subscriberLocale = Integer.parseInt(st.nextToken());
      this.grb = new GameResourceBundle(this.subscriberLocale);
      this.firstTime = Long.parseLong(st.nextToken());
      if (this.firstTime <= 0) this.firstTime = System.currentTimeMillis();
      this.lastTime = System.currentTimeMillis();
      this.totalGamesCount = Long.parseLong(st.nextToken());
      this.winGamesCount = Long.parseLong(st.nextToken());
      this.currScore = new GameScore();
      int time = Integer.parseInt(st.nextToken());
      int moves = Integer.parseInt(st.nextToken());
      this.bestScore = (time >= 0 && moves >= 0) ? new Score(time, moves) : null;
      this.bestScoreSet = false;
    } catch (Exception exc) {
      throw new Exception(GAME_LOAD_ERROR + PARSE_FAILED + "details=" + exc.getMessage());
    }
  }

  private void loadGameSess(String data) throws Exception
  {
    final String GAME_LOAD_ERROR = "Game load error (sess): ";
    final String PARSE_FAILED = "Failed to parse game data ";

    newCells();
    if (data == null)
      throw new Exception(GAME_LOAD_ERROR + "Game data is null");

    try {
      StringTokenizer st = new StringTokenizer(data, ",");
      for (int y=0,i=0; y<4; y++) {
        for (int x=0; x<4; x++, i++) {
          if (!st.hasMoreTokens())
            throw new Exception("(cells)");
          byte b = Byte.parseByte(st.nextToken());
          if (b<0 || b>15)
            throw new Exception("(cell range[0..15])");
          this.cells[y][x]=b; if (b == 0) this.empty = i;
        }
      }

      this.lastTime = Long.parseLong(st.nextToken());
      if (this.lastTime <= 0) this.lastTime = System.currentTimeMillis();
      int time = Integer.parseInt(st.nextToken());
      int moves = Integer.parseInt(st.nextToken());
      this.currScore = new GameScore(time, moves);
      this.bestScoreSet = false;
      this.gameStarted = (st.hasMoreTokens() ? Integer.parseInt(st.nextToken()) > 0 : !isWin());
    } catch (Exception exc) {
      throw new Exception(GAME_LOAD_ERROR + PARSE_FAILED + "details=" + exc.getMessage());
    }
  }

  public String getPersData()
  {
    final StringBuilder sb = new StringBuilder(64);
    sb.append(this.subscriberLocale); sb.append(',');
    sb.append(this.firstTime); sb.append(',');
    sb.append(this.totalGamesCount); sb.append(',');
    sb.append(this.winGamesCount); sb.append(',');
    sb.append(this.bestScore != null ? bestScore.getTime()  : -1); sb.append(',');
    sb.append(this.bestScore != null ? bestScore.getMoves() : -1);
    return sb.toString();
  }

  public String getSessData()
  {
    final StringBuilder sb = new StringBuilder(64);
    for (int y=0,i=0; y<4; y++) {
      for (int x=0; x<4; x++,i++) {
        byte b = cells[y][x];
        if( b<0 || b>15 ) return null;
        sb.append(b); sb.append(',');
      }
    }
    sb.append(this.lastTime); sb.append(',');
    sb.append(this.currScore.getTime()); sb.append(',');
    sb.append(this.currScore.getMoves()); sb.append(',');
    sb.append(this.gameStarted ? 1:0);
    return sb.toString();
  }

  public String getStatus() {
    final StringBuilder sb = new StringBuilder(64);
    sb.append(grb.getString("msg.st.status"));sb.append(' ');
    sb.append(grb.getString("msg.st.moves")); sb.append(' '); sb.append(currScore.getMoves());
    sb.append(','); sb.append(' '); sb.append(grb.getString("msg.st.time")); sb.append(' ');
    sb.append(getTimeString(currScore));
    return sb.toString();
  }

  public boolean isWin() {
    if (cells == null) return false;
    for (int y=0,i=0; y<4; y++) {
      for (int x=0; x<4; x++) {
        if (cells[y][x] == 0) continue;
        if (cells[y][x] != ++i) return false;
      }
    }
    this.gameStarted = false;
    return true;
  }
  public boolean hasBestScore() {
    return bestScore != null;
  }
  private boolean isBestScore() {
    return bestScore == null || currScore.compareTo(bestScore) < 0;
  }
  public boolean isBestScoreSet() {
    return bestScoreSet;
  }
  public String getBestScoreTime() {
    return getTimeString(bestScore);
  }
  public int getBestScoreMoves() { return bestScore.moves; }

  public String getTimeString(Score score) {
    final StringBuilder sb = new StringBuilder(64);
    int time = score.getTime();
    int val = time / (3600 * 24);
    if (val > 0) { sb.append(val); sb.append(' '); sb.append(grb.getString("msg.st.days")); sb.append(' '); }
    val = (time / 3600) % 24; if (val < 10) sb.append('0'); sb.append(val); sb.append(':');
    val = (time / 60) % 60; if (val < 10) sb.append('0'); sb.append(val); sb.append(':');
    val = time % 60; if (val < 10) sb.append('0'); sb.append(val);
    return sb.toString();
  }

  public boolean isStarted() {
    return this.gameStarted;
  }
  public boolean isInProgress() {
    return !isWin() && this.gameStarted && lastTime > 0 && ((System.currentTimeMillis() - lastTime) < GameStore.getMaxGameActivityTime()*1000);
  }

  private byte get(int index) {
    if (index < 0 || index > 15) return -1;
    return cells[index/4][index%4];
  }
  private boolean set(int index, byte value) {
    if (index < 0 || index > 15) return false;
    cells[index/4][index%4] = value;
    return true;
  }
  private void incMoves()
  {
    this.currScore.incMoves();
    this.lastTime = System.currentTimeMillis();
    this.bestScoreSet = false;
    if (isWin()) {
      this.gameStarted = false;
      this.winGamesCount++;
      if (isBestScore()) {
        this.bestScore = new Score(currScore);
        this.bestScoreSet = true;
      }
      GameStore.storePersData(this);
    }
    GameStore.storeSessData(this);
  }
  public boolean move(int index)
  {
    if (index < 0 || empty == index || index > 15) return false;

    if (empty/4 == index/4)       // empty cell & index in the same row
    {
      while (empty > index) {     // shift cells right (empty cell moves to the left)
        set(empty, get(empty - 1)); empty--; set(empty, (byte)0);
      }
      while (empty < index) {     // shift cells left (empty moves to the right)
        set(empty, get(empty + 1)); empty++; set(empty, (byte)0);
      }
      incMoves();
      return true;
    }
    else if (empty%4 == index%4)  // empty cell & index in the same column
    {
      while (empty > index) {     // shift cells down (empty cell moves up)
        set(empty, get(empty - 4)); empty -= 4; set(empty, (byte) 0);
      }
      while (empty < index) {     // shift cells up (empty cell moves down)
        set(empty, get(empty + 4)); empty += 4; set(empty, (byte) 0);
      }
      incMoves();
      return true;
    }
    return false;
  }

}


src/mobi/eyeline/bots/game15/GameResourceBundle.java
package mobi.eyeline.bots.game15;

import java.util.Locale;
import java.util.ResourceBundle;

public class GameResourceBundle
{
  private ResourceBundle bundle;

  public  final static int RU_RU_LOCALE   = 0;
  public  final static int EN_US_LOCALE   = 1;
  public  final static int DEFAULT_LOCALE = RU_RU_LOCALE;

  private final Locale[] locales = new Locale[] { new Locale("ru", "ru"), new Locale("en", "us")};
  private int currentLocale = DEFAULT_LOCALE;

  private ResourceBundle getBundle(int locale) {
    return ResourceBundle.getBundle("game15", locales[locale]);
  }
  private void setLocale(int locale) {
    if (locale < 0 || locale >= locales.length) locale = DEFAULT_LOCALE;
    this.currentLocale = locale; this.bundle = getBundle(currentLocale);
  }

  public GameResourceBundle() {
    this(DEFAULT_LOCALE);
  }
  public GameResourceBundle(int locale) {
    setLocale(locale);
  }

  private int getNextLocaleIndex() {
    return (currentLocale+1)%locales.length;
  }
  public synchronized void switchLocale() {
    setLocale(getNextLocaleIndex());
  }
  private String getString(String key, ResourceBundle bundle) {
    String value = bundle.getString(key);
    try {
      return new String(value.getBytes("ISO-8859-1"), "UTF-8");
    } catch (Exception e) {
      return value;
    }
  }
  public String getString(String key){
    return getString(key, bundle);
  }
  public String getNextString(String key){
    ResourceBundle bundle = getBundle(getNextLocaleIndex());
    return getString(key, bundle);
  }

  protected synchronized Locale getCurrentLocale() {
    return locales[currentLocale];
  }

}


src/mobi/eyeline/bots/game15/GameSessionListener.java
package mobi.eyeline.bots.game15;

import org.apache.log4j.PropertyConfigurator;

import javax.servlet.ServletContextEvent;
import javax.servlet.ServletContextListener;
import javax.servlet.http.HttpSession;
import javax.servlet.http.HttpSessionEvent;
import javax.servlet.http.HttpSessionListener;

import java.io.File;
import java.util.Date;
import java.util.concurrent.TimeUnit;

public class GameSessionListener implements HttpSessionListener, ServletContextListener
{
  private String getTime() {
    return new Date(System.currentTimeMillis()).toString();
  }

  @Override
  public void sessionCreated(HttpSessionEvent event) {
    HttpSession session = event.getSession();
    System.out.print(getTime() + " (session) Created:");
    System.out.println("ID=" + session.getId() + " MaxInactiveInterval=" + session.getMaxInactiveInterval());
  }

  @Override
  public void sessionDestroyed(HttpSessionEvent event) {
    HttpSession session = event.getSession();
    System.out.println(getTime() + " (session) Destroyed:ID=" + session.getId());
    GameStore.finalizeGame(session.getId());
  }

  @Override
  public void contextInitialized(ServletContextEvent servletContextEvent)
  {
    final String configDir = System.getProperty("game15.config_dir");
    if (configDir == null)
      throw new RuntimeException("Failed to obtain Game15 config directory");

    final File log4jprops = new File(configDir, "log4j.properties");
    System.out.println("Log4j conf file: " + log4jprops.getAbsolutePath() +
                       ", exists: " + log4jprops.exists());

    PropertyConfigurator.configureAndWatch(log4jprops.getAbsolutePath(), TimeUnit.MINUTES.toMillis(1));

    GameStore.init(configDir);
  }

  @Override
  public void contextDestroyed(ServletContextEvent servletContextEvent) {

  }
}


src/mobi/eyeline/bots/game15/GameStore.java
package mobi.eyeline.bots.game15;

import static mobi.eyeline.utils.restclient.web.RestClient.*;

import mobi.eyeline.utils.restclient.web.RestClient;
import org.apache.log4j.Logger;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.*;

public class GameStore
{
  private final static Logger logger = Logger.getLogger(GameStore.class);
  private final static Object storeSync = new Object();

  private static HashMap<String, Game> gamesBySubscriber = null;
  private static HashMap<String, Game> getGamesBySubscriber() {
    if (gamesBySubscriber == null) { gamesBySubscriber = new HashMap<>(); }
    return gamesBySubscriber;
  }
  private static HashMap<String, Game> gamesBySession = null;
  private static HashMap<String, Game> getGamesBySession() {
    if (gamesBySession == null) { gamesBySession = new HashMap<>(); }
    return gamesBySession;
  }

  private static boolean  configured = false;
  private static int      maxGameActivityTime = 0;
  private static String   ANOTHER_GAMES_URL;
  private static String   REST_PERS_API_ROOT;
  private static String   REST_SESS_API_ROOT;
  private static String   GAME_VAR_BASE;
  private static String   PERS_GAME_VAR;
  private static String   SESS_GAME_VAR;

  public static boolean init(String configDir)
  {
    synchronized (storeSync) {
      if (!configured) {
        ResourceBundle bundle = null;
        File configFile = new File(configDir, "config.properties");
        try (FileInputStream fis = new FileInputStream(configFile)) {
          bundle = new PropertyResourceBundle(fis);
        } catch (IOException e) {
          logger.warn("Failed to load Game15 config", e);
          return false;
        }
        maxGameActivityTime   = Integer.parseInt(bundle.getString("game.activity.time"));
        ANOTHER_GAMES_URL     = bundle.getString("another.games.url");
        REST_PERS_API_ROOT    = bundle.getString("api.root.pers");
        REST_SESS_API_ROOT    = bundle.getString("api.root.sess");
        GAME_VAR_BASE         = bundle.getString("game.var.base");
        PERS_GAME_VAR = GAME_VAR_BASE + ".pers";
        SESS_GAME_VAR = GAME_VAR_BASE + ".sess";
        configured = true;
      }
    }
    return true;
  }
  public static int getMaxGameActivityTime() {
    return maxGameActivityTime;
  }
  public static String getAnotherGamesUrl() {
    return ANOTHER_GAMES_URL;
  }
  private static boolean storeGameData(String apiRoot, String varName, String subscriber, String data) {
    try {
      new RestClient().json(apiRoot + "profile/" + subscriber + "/" + varName, post(RestClient.content(data)));
    } catch (IOException exc) {
      logger.warn("Store game for '" + subscriber + "', var=" + varName + ", data='" + data + "'. Failed. Details: " + exc.getMessage());
      return false;
    }
    if (logger.isDebugEnabled()) {
      logger.debug("Store game for '" + subscriber + "', var=" + varName + ", data='" + data + "'. Ok");
    }
    return true;
  }
  public static boolean storePersData(Game game) {
    return storeGameData(REST_PERS_API_ROOT, PERS_GAME_VAR, game.getSubscriber(), game.getPersData());
  }
  public static boolean storeSessData(Game game) {
    return storeGameData(REST_SESS_API_ROOT, SESS_GAME_VAR, game.getSubscriber(), game.getSessData());
  }

  private static String loadGameData(String apiRoot, String varName, String subscriber) {
    String data;
    try {
      data = (String)((new RestClient()).json(apiRoot + "profile/" + subscriber + "/" + varName).object().get("value"));
    } catch (Exception exc) {
      logger.warn("Load game for '" + subscriber + "', var=" + varName + ". Failed. Details: " + exc.getMessage());
      return null;
    }

    if (logger.isDebugEnabled()) {
      logger.debug("Load game for '" + subscriber + "', var=" + varName + ", data='" + data + "'. Ok");
    }
    return data;
  }
  private static String loadPersData(String subscriber) {
    return loadGameData(REST_PERS_API_ROOT, PERS_GAME_VAR, subscriber);
  }
  private static String loadSessData(String subscriber) {
    return loadGameData(REST_SESS_API_ROOT, SESS_GAME_VAR, subscriber);
  }

  public static Game newGame(String subscriber, String sessionId, String gp_lang) {
    Game game = new Game(subscriber);
    game.setGpLang(gp_lang);
    if (logger.isDebugEnabled()) {
      logger.debug("New game for '" + subscriber + "' SID=" + sessionId);
    }
    synchronized (storeSync) {
      Game old = getGamesBySubscriber().put(subscriber, game);
      if (old != null) {
        if (logger.isDebugEnabled()) {
          logger.warn("New game for '" + subscriber + "' SID=" + sessionId + ". Found old game by subscriber. Replaced");
        }
      }
      old = getGamesBySession().put(sessionId, game);
      if (old != null) {
        if (logger.isDebugEnabled()) {
          logger.warn("New game for '" + subscriber + "' SID=" + sessionId + ". Found old game by sessionId. Replaced");
        }
      }
    }
    storePersData(game);
    return game;
  }

  public static Game getGame(String subscriber, String sessionId) {
    Game game;
    synchronized (storeSync) {
      game = getGamesBySubscriber().get(subscriber);
      if (game != null) {
        if (logger.isDebugEnabled()) {
          logger.debug("Get game for '" + subscriber + "' SID=" + sessionId + ". Got by subscriber");
        }
        return game;
      }
    }
    if (logger.isDebugEnabled()) {
      logger.debug("Get game for '" + subscriber + "' SID=" + sessionId + ". Loading...");
    }
    game = loadGame(subscriber);
    if (game != null) {
      if (logger.isDebugEnabled()) {
        logger.debug("Get game for '" + subscriber + "' SID=" + sessionId + ". Loaded");
      }
      synchronized (storeSync) {
        getGamesBySubscriber().put(game.getSubscriber(), game);
        getGamesBySession().put(sessionId, game);
      }
    }
    return game;
  }

  public static boolean finalizeGame(String sessionId) {
    Game game;
    synchronized (storeSync) {
      game = getGamesBySession().remove(sessionId);
      if (game == null) {
        if (logger.isDebugEnabled()) {
          logger.debug("Fin game for SID=" + sessionId + ". Not found");
        }
        return false;
      }
      getGamesBySubscriber().remove(game.getSubscriber());
      if (logger.isDebugEnabled()) {
        logger.debug("Fin game for '" + game.getSubscriber() + "' SID=" + sessionId + ". Ok");
      }
    }
    return (storePersData(game) && storeSessData(game));
  }

  private static Game loadGame(final String subscriber) {
    Game game;
    try {
      game = new Game(subscriber, loadPersData(subscriber), loadSessData(subscriber));
    } catch (Exception exc) {
      logger.warn("Load game for '" + subscriber + "'. Failed. Details: " + exc.getMessage());
      return null;
    }
    if (logger.isDebugEnabled()) {
      logger.debug("Load game for '" + subscriber + "'. Ok");
    }
    return game;
  }

/*
  private static ClientConfig pvssClientConfig = null;
  private static SetCommand pvssSetCommand = null;
  private static GetCommand pvssGetCommand = null;
  private static Protocol pvssProtocol = null;
  private static Client pvssClient = null;
  private static Client getPvssClient() {
    synchronized (storeSync) {
      if (pvssClient == null) {
        pvssClientConfig = new ClientConfig(); // TODO: init from properties file
        pvssProtocol = new PvapProtocol();
        pvssClient = new ClientCore(pvssClientConfig, pvssProtocol);
        pvssSetCommand = new SetCommand();
        pvssSetCommand.setVarName(PVSS_VAR_NAME);
        pvssSetCommand.getProperty().setTimePolicy(new InfinitTimePolicy());
        pvssGetCommand = new GetCommand();
        pvssGetCommand.setVarName(PVSS_VAR_NAME);
        try {
          pvssClient.startup();
        } catch (PvssException e) {
          logger.error(e.getMessage());
        }
      }
    }
    return pvssClient;
  }

  private final static String PVSS_VAR_NAME = "game15";
  private static boolean storePVSSGame(Game game)
  {
    final String subscriber = game.getSubscriber();
    final String gameData = game.toString();
    if (logger.isDebugEnabled()) {
      logger.debug("storeGame for '" + subscriber + "'");
    }
    try {
      pvssSetCommand.setStringValue(gameData);
      final ProfileRequest pvssReq = new ProfileRequest(pvssSetCommand);
      pvssReq.getProfileKey().setAbonentKey(subscriber);
      final Response pvssResp = getPvssClient().processRequestSync(pvssReq);
      if (logger.isDebugEnabled()) {
        logger.debug("storeGame for '" + subscriber + "' PVSS Response: " + pvssResp);
      }
    } catch (PvssException e) {
      logger.error("storeGame for '" + subscriber + "' PVSS Error: " + e.getMessage());
      return false;
    }

    if (logger.isDebugEnabled()) {
      logger.debug("storeGame for '" + game.getSubscriber() + "'. Stored");
    }
    return true;
  }

  private final static class PVSSResponseVisitor implements ResponseVisitor
  {
    private final String subscriber;
    private String gameData = null;

    private PVSSResponseVisitor(String subscriber) {
      this.subscriber = subscriber;
    }
    public String getGameData() {
      return gameData;
    }

    @Override
    public boolean visitErrResponse(ErrorResponse resp) throws PvssException {
      if (logger.isDebugEnabled()) {
        logger.debug("PVSS for '" + subscriber + "' Err Response: " + resp);
      }
      return true;
    }
    @Override
    public boolean visitProfileResponse(ProfileResponse resp) throws PvssException {
      if (logger.isDebugEnabled()) {
        logger.debug("PVSS for '" + subscriber + "' Ok Response: " + resp);
      }
      gameData = ((GetResponse)resp.getResponse()).getStringValue();
      return true;
    }
    @Override
    public boolean visitPingResponse(PingResponse resp) throws PvssException {
      return false;
    }
    @Override
    public boolean visitAuthResponse(AuthResponse resp) throws PvssException {
      return false;
    }
  }
  private static Game loadPVSSGame(final String subscriber)
  {
    if (logger.isDebugEnabled()) {
      logger.debug("loadGame for '" + subscriber + "'");
    }

    Game game = null;
    String gameData = null; // "0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,0,-1,-1,0,0,0,0,0,0";
    try {
      final ProfileRequest pvssReq = new ProfileRequest(pvssGetCommand);
      pvssReq.getProfileKey().setAbonentKey(subscriber);
      final Response pvssResp = getPvssClient().processRequestSync(pvssReq);
      if (logger.isDebugEnabled()) {
        logger.debug("loadGame for '" + subscriber + "' PVSS Response: " + pvssResp);
      }
      final PVSSResponseVisitor grv = new PVSSResponseVisitor(subscriber);
      pvssResp.visit(grv);
      gameData = grv.getGameData();
    } catch (PvssException e) {
      logger.error("loadGame for '" + subscriber + "' PVSS Error: " + e.getMessage());
    }

    try {
      game = new Game(subscriber, gameData);
    } catch (Exception e) {
      logger.warn("loadGame for '" + subscriber + "'. Failed. Details: " + e.getMessage());
      return null;
    }

    if (logger.isDebugEnabled()) {
      logger.debug("loadGame for '" + subscriber + "'. Loaded");
    }
    return game;
  }
*/
}
